pkg/engine/plugins.go

```diff
@@ -56,6 +56,33 @@ func (p pluginSet) Union(other pluginSet) pluginSet {
 	return newSet
 }
 
+// Removes less specific entries.
+//
+// For example, the plugin aws would be removed if there was an already existing plugin
+// aws-5.4.0.
+func (p pluginSet) Deduplicate() pluginSet {
+	existing := map[string]workspace.PluginInfo{}
+	newSet := newPluginSet()
+	add := func(p workspace.PluginInfo) {
+		prev, ok := existing[p.Name]
+		if ok && p.Version == nil && p.PluginDownloadURL == "" {
+			// no new information
+			return
+		}
+
+		if prev.Version == nil && prev.PluginDownloadURL == "" {
+			// New plugin is more specific then the old one
```

Comment:
What if the old just had version set and the new just has downloadURL set? What if we get two with version set but to different versions (5.0 and 6.0 for example). It's not clear there's a correct total ordering here?

Reply:
If either `pluginDownloadURL`, `Version` or both are set we consider the plugin fully specified and keep it. It is ok to keep `pkg v1.2.3` `pkg v2.3.4` and `pkg example.com` in a single set. What we don't want to do is keep `pkg` in that same set since there are more specific versions used. In general, there will be a `pky vx.y.y` in the plugin set because the user depended on a language package `pulumi-pkg` with version `x.y.z`.